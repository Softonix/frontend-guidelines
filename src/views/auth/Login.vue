<template>
  <div class="h-full flex items-center justify-center">
    <el-card class="w-[450px] max-w-full m-auto">
      <el-form
        ref="formRef"
        :model="formModel"
        :rules="formRules"
        label-position="top"
      >
        <el-form-item label="Email" prop="email">
          <el-input v-model="formModel.email" />
        </el-form-item>

        <el-form-item label="Password" prop="password">
          <el-input v-model="formModel.password" />
        </el-form-item>

        <div>
          <el-button
            :type="'primary'"
            data-testid="loginBtn"
            @click="submitForm(formRef.validate)"
          >
            {{ t('auth.login') }}
          </el-button>

          <el-button
            data-testid="resetBtn"
            @click="resetForm(formRef.resetFields)"
          >
            {{ t('auth.reset') }}
          </el-button>

          <el-button
            data-testid="backToHomeBtn"
            @click="$router.push({ name: $routeNames.rootPage })"
          >
            {{ t('auth.backToHome') }}
          </el-button>
        </div>

        <div class="mt-3" data-testid="feature-flag">
          <FF name="FF-SX-1924-LOGIN-FLAG-EXAMPLE" show-badge>
            <el-button>Feature Flag Enabled</el-button>
          </FF>
        </div>
      </el-form>
    </el-card>
  </div>
</template>

<script lang="ts" setup>
const { submitForm, resetForm } = useForm()
const { t } = useI18n()

const formRef = useElFormRef()
const formModel = reactive({
  email: '',
  password: ''
})

const formRules = useElFormRules({
  email: [
    useRequiredRule(),
    useEmailRule()
  ],
  password: [
    useRequiredRule(),
    useMinLenRule(5)
  ]
})
</script>
