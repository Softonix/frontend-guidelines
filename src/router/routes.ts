import type { RouteRecordRaw } from 'vue-router'

import { routeNames } from './route-names'

import { authRoutes } from '@/views/auth/auth.routes'
import { exampleViewRoutes } from '@/views/example-view/example-view.routes'

import DefaultLayout from '@/layouts/DefaultLayout.vue'

const defaultLayoutRoutes: RouteRecordRaw = {
  path: '/',
  name: routeNames.rootPage,
  redirect: { name: routeNames.exampleGeneral },
  component: DefaultLayout,
  children: [
    // list of views that use default layout
    ...exampleViewRoutes
  ]
}

const routes: Array<RouteRecordRaw> = [
  {
    path: '/:pathMatch(.*)*',
    redirect: '/'
  },

  authRoutes,
  defaultLayoutRoutes,
  {
    path: '/react-vs-vue',
    component: DefaultLayout,
    children: [{ path: '', name: 'react-vs-vue', component: () => import('@/views/react-vs-vue/index.vue') }]

  }
]

export {
  routes,
  defaultLayoutRoutes
}
