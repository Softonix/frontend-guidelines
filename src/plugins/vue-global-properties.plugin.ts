import type { Plugin } from 'vue'
import { routeNames } from '@/router/route-names'

export const globalProperties = {
  $routeNames: routeNames
}

export const VueGlobalPropertiesPlugin: Plugin = {
  install: (app) => {
    for (const i in globalProperties) {
      app.config.globalProperties[i] = globalProperties[i as keyof typeof globalProperties]
    }
  }
}
