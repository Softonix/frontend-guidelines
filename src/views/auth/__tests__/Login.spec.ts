import { it, describe, expect, beforeEach, afterEach, vi } from 'vitest'
import { render, fireEvent, screen, waitFor, cleanup } from '@testing-library/vue'
import { faker } from '@faker-js/faker'
import { i18n } from '@/plugins'
import { routeNames } from '@/router/route-names'
import Login from '../Login.vue'

const mockRouter = { push: vi.fn() }
const mockRoute = { name: routeNames.rootPage }

beforeEach(() => {
  render(Login, {
    global: {
      plugins: [i18n],
      mocks: {
        $routeNames: routeNames,
        $route: mockRoute,
        $router: mockRouter
      }
    }
  })
})

afterEach(() => {
  cleanup()
})

describe('Login component', () => {
  describe('login flow tests', () => {
    it('should validate required fields', async () => {
      const submitButton = screen.getByTestId('loginBtn')

      await fireEvent.click(submitButton)

      await waitFor(() => {
        const emailError = screen.queryByText('Required field', { selector: 'div' })
        const passError = screen.queryByText('Required field', { selector: 'div' })

        expect(emailError).not.toBeNull()
        expect(passError).not.toBeNull()
      })
    })
  })

  describe('check email field validations', () => {
    it('should validate email field', async () => {
      const emailField = screen.getByLabelText('Email') as HTMLInputElement

      const testEmail = faker.internet.email()

      await fireEvent.update(emailField, testEmail)

      await waitFor(() => {
        expect(emailField.value).toBe(testEmail)
      })
    })

    it('should show an error message if invalid email provided', async () => {
      const emailField = screen.queryByLabelText('Email') as HTMLInputElement
      const invalidEmail = 'invalid'
      const errorMessage = 'Invalid email'

      await fireEvent.update(emailField, invalidEmail)

      await waitFor(() => {
        const emailError = screen.findByText(errorMessage, { selector: 'div' })
        expect(emailError).not.toBeNull()
      })
    })
  })

  describe('check password field validations', () => {
    it('should validate password field', async () => {
      const emailField = screen.getByLabelText('Password') as HTMLInputElement

      const testEmail = faker.internet.password()

      await fireEvent.update(emailField, testEmail)

      await waitFor(() => {
        expect(emailField.value).toBe(testEmail)
      })
    })

    it('should show an error message if invalid password provided', async () => {
      const passwordField = screen.queryByLabelText('Password') as HTMLInputElement
      const invalidPassword = 'sdc'
      const errorMessage = 'Min length 5'

      await fireEvent.update(passwordField, invalidPassword)

      await waitFor(() => {
        const emailError = screen.findByText(errorMessage, { selector: 'div' })
        expect(emailError).not.toBeNull()
      })
    })
  })

  describe('check form buttons', () => {
    it('should reset form fields by clicking on reset button', async () => {
      const emailField = screen.getByLabelText('Email') as HTMLInputElement
      const passwordField = screen.getByLabelText('Password') as HTMLInputElement
      const resetButton = screen.getByTestId('resetBtn')

      const fakeEmail = faker.internet.email()
      const fakePassword = faker.internet.password()

      await fireEvent.update(emailField, fakeEmail)
      await fireEvent.update(passwordField, fakePassword)
      await fireEvent.click(resetButton)

      await waitFor(() => {
        expect(emailField.value).toBe('')
        expect(passwordField.value).toBe('')
      })
    })

    it('should back to the main page when back button was clicked', async () => {
      const backButton = screen.getByTestId('backToHomeBtn')

      await fireEvent.click(backButton)

      expect(mockRouter.push).toHaveBeenCalledOnce()
      expect(mockRouter.push).toHaveBeenCalledWith({ name: routeNames.rootPage })
    })
  })
})
