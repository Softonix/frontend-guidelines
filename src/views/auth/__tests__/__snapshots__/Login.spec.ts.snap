// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login component > should create a Login component snapshot 1`] = `
<div
  class="h-full flex items-center justify-center"
  data-testid="login-root"
>
  <div
    class="el-card is-always-shadow w-[450px] max-w-full m-auto"
  >
    <!--v-if-->
    <div
      class="el-card__body"
      style=""
    >
      
      <form
        class="el-form el-form--default el-form--label-top"
      >
        
        <div
          class="el-form-item is-required asterisk-left"
        >
          
          
          <label
            class="el-form-item__label"
            for="el-id-6357-34"
            id="el-id-6357-32"
          >
            
            Email
            
          </label>
          
          
          <div
            class="el-form-item__content"
          >
            
            <div
              class="el-input"
            >
              <!-- input -->
              
              <!-- prepend slot -->
              <!--v-if-->
              <div
                class="el-input__wrapper"
              >
                <!-- prefix slot -->
                <!--v-if-->
                <input
                  autocomplete="off"
                  class="el-input__inner"
                  id="el-id-6357-34"
                  tabindex="0"
                  type="text"
                />
                <!-- suffix slot -->
                <!--v-if-->
              </div>
              <!-- append slot -->
              <!--v-if-->
              
            </div>
            
            <transition-group-stub
              appear="false"
              css="true"
              name="el-zoom-in-top"
              persisted="false"
            >
              <!--v-if-->
            </transition-group-stub>
          </div>
        </div>
        <div
          class="el-form-item is-required asterisk-left"
        >
          
          
          <label
            class="el-form-item__label"
            for="el-id-6357-35"
            id="el-id-6357-33"
          >
            
            Password
            
          </label>
          
          
          <div
            class="el-form-item__content"
          >
            
            <div
              class="el-input"
            >
              <!-- input -->
              
              <!-- prepend slot -->
              <!--v-if-->
              <div
                class="el-input__wrapper"
              >
                <!-- prefix slot -->
                <!--v-if-->
                <input
                  autocomplete="off"
                  class="el-input__inner"
                  id="el-id-6357-35"
                  tabindex="0"
                  type="text"
                />
                <!-- suffix slot -->
                <!--v-if-->
              </div>
              <!-- append slot -->
              <!--v-if-->
              
            </div>
            
            <transition-group-stub
              appear="false"
              css="true"
              name="el-zoom-in-top"
              persisted="false"
            >
              <!--v-if-->
            </transition-group-stub>
          </div>
        </div>
        <div>
          <button
            ariadisabled="false"
            class="el-button el-button--primary"
            data-testid="loginBtn"
            type="button"
          >
            <!--v-if-->
            <span
              class=""
            >
              
              Login
              
            </span>
          </button>
          <button
            ariadisabled="false"
            class="el-button"
            data-testid="resetBtn"
            type="button"
          >
            <!--v-if-->
            <span
              class=""
            >
              
              Reset
              
            </span>
          </button>
          <button
            ariadisabled="false"
            class="el-button"
            data-testid="backToHomeBtn"
            type="button"
          >
            <!--v-if-->
            <span
              class=""
            >
              
              Back to home page
              
            </span>
          </button>
        </div>
        <div
          class="mt-3"
          data-testid="feature-flag"
        >
          <!--v-if-->
        </div>
        
      </form>
      
    </div>
  </div>
</div>
`;
